compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaLexer.java

```diff
@@ -1306,8 +1306,17 @@ private STToken processExclamationMarkOperator() {
     }
 
     private boolean isNotIsToken() {
-        return (reader.peek() == 'i' && reader.peek(1) == 's') &&
-                !(isIdentifierFollowingChar(reader.peek(2)) || reader.peek(2) == LexerTerminals.BACKSLASH);
+        int offset = reader.getOffset();
+        reader.mark();
+        while (isIdentifierFollowingChar(reader.peek())) {
+            reader.advance();
+        }
+
+        if (reader.getMarkedChars().equals(LexerTerminals.IS) && !(reader.peek() == LexerTerminals.BACKSLASH)) {
+            return true;
+        }
+        reader.reset(offset);
```

Comment:
Incrementing the reader here and then reset is sort of hacky.
We can advance to the reader only after checking the incoming char sequence.
Maybe we can do something like this?

```java
    private STToken processAndGetExclamationMarkOrNotIs() {
        StringBuilder stringBuilder = new StringBuilder();
        int lookaheadCount = 0;
        int lookaheadChar = reader.peek(lookaheadCount);
        while (isIdentifierFollowingChar(lookaheadChar)) {
            stringBuilder.append((char) lookaheadChar);
            lookaheadCount++;
            lookaheadChar = reader.peek(lookaheadCount);
        }

        if (stringBuilder.toString().equals(LexerTerminals.IS) && !(lookaheadChar == LexerTerminals.BACKSLASH)) {
            reader.advance(lookaheadCount);
            return getSyntaxToken(SyntaxKind.NOT_IS_KEYWORD);
        } else {
            return getSyntaxToken(SyntaxKind.EXCLAMATION_MARK_TOKEN);
        }
    }
```

Reply:
+1